The type hint for the filename parameter to open_file is incomplete

`click.open_file(filename, …)` happily accepts a `pathlib.Path` object as the filename; it [only passes the value along][open-file-src] to either `click.utils.LazyFile(filename, …)` or `click._compat.open_stream(filename, …)`, both of which accept pathlikes. However, `open_file`'s type hints indicate that `filename` can only be a string, which causes type-checking errors when feeding it a `Path`.

[open-file-src]: https://github.com/pallets/click/blob/874ca2bc1c30d93a4ac6e36a15ed685eafe89097/src/click/utils.py#L391-L401

``` python
from pathlib import Path

import click

filename = Path("/some/path")

# Mypy gives the following error on the line with the call to open_file:
#
# error: Argument 1 to "open_file" has incompatible type "Path"; expected "str"  [arg-type]

with click.open_file(filename) as handle:
    pass
```

Both `LazyFile` and `open_stream` give their filename parameter the type `types.Union[str, os.PathLike[str]]`, so that should probably be the type accepted by `open_file`, too.

Environment:

- Python version: 3.12.1
- Click version: 8.1.7

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,5 +1,12 @@
 .. currentmodule:: click
 
+Version 8.1.8
+-------------
+
+Unreleased
+
+-   Fix an issue with type hints for ``click.open_file()``. :issue:`2717`
+
 Version 8.1.7
 -------------
 

--- a/src/click/utils.py
+++ b/src/click/utils.py
@@ -353,7 +353,7 @@ def get_text_stream(
 
 
 def open_file(
-    filename: str,
+    filename: t.Union[str, "os.PathLike[str]"],
     mode: str = "r",
     encoding: t.Optional[str] = None,
     errors: t.Optional[str] = "strict",
@@ -374,7 +374,7 @@ def open_file(
         with open_file(filename) as f:
             ...
 
-    :param filename: The name of the file to open, or ``'-'`` for
+    :param filename: The name or Path of the file to open, or ``'-'`` for
         ``stdin``/``stdout``.
     :param mode: The mode in which to open the file.
     :param encoding: The encoding to decode or encode a file opened in

